src/components/Viewport.ts

```diff
@@ -546,26 +653,66 @@ export default class Viewport {
     const panelManager = this.panelManager;
 
     // Restore index
-    panelManager.clear();
-    cameraElement.innerHTML = status.panels.map(panel => panel.html).join("");
-
-    this.createPanels();
-
-    // Reset panel index
-    panelManager.originalPanels().forEach((panel, idx) => {
-      panel.setIndex(panels[idx].index);
+    cameraElement.innerHTML = panels.map(panel => panel.html).join("");
+
+    // Create panels first
+    this.refreshPanels();
+    const createdPanels = panelManager.originalPanels();
+
+    // ...then order it by its index
+    const orderedPanels: Panel[] = [];
+    panels.forEach((panel, idx) => {
+      const createdPanel = createdPanels[idx];
+      createdPanel.setIndex(panel.index);
+      orderedPanels[panel.index] = createdPanel;
     });
-
-    this.currentPanel = panelManager.get(status.index)
-      || panelManager.get(defaultIndex)
-      || panelManager.firstPanel();
+    panelManager.replacePanels(orderedPanels, []);
+    panelManager.setCloneCount(0); // No clones at this point
+
+    const panelCount = panelManager.getPanelCount();
+    if (panelCount > 0) {
+      this.currentPanel = panelManager.get(status.index)
+        || panelManager.get(defaultIndex)
+        || panelManager.firstPanel();
+      this.nearestPanel = this.currentPanel;
+    } else {
+      this.currentPanel = undefined;
+      this.nearestPanel = undefined;
+    }
+    this.visiblePanels = orderedPanels.filter(panel => Boolean(panel));
 
     this.resize();
 
     this.axes.setTo({ flick: status.position }, 0);
     this.moveCamera(status.position);
   }
 
+  public calcVisiblePanels(): Panel[] {
```

Comment:
Nope, getVisiblePanels just feels like
```js
public getVisiblePanels() {
  return this.visiblePanels
}
```

Reply:
But Flicking's getVisiblePanels already use calcVisiblePanels.

Reply:
But I don't think Flicking's `getVisiblePanels()` must meet Viewport's `getVisiblePanels()`.
Because Viewport's `visiblePanels` are only calculated when `renderOnlyVisible` is on.
If `renderOnlyVisible` is off, it must be calculated with outset progress rather than got.